src/visualizers-workbench/P5VisualizerTemplate.ts

```diff
@@ -0,0 +1,174 @@
+import {P5Visualizer} from '../visualizers/P5Visualizer'
+import {VisualizerExportModule} from '@/visualizers/VisualizerInterface'
+
+/** md
+# p5 Visualizer Template
+
+[image should go here]
+
+This template can be used as a starting point for designing a p5 visualizer. You
+can use it to step back and forth through a sequence, showing each element as a
+written number.
+
+## Parameters
+
+- Step size
+
+## Interactions
+
+- Use the left and right arrow keys to step
+**/
+
+class P5VisualizerTemplate extends P5Visualizer {
+    // the name that appears in bundle card titles
+    static visualizationName = 'p5 Visualizer Template'
+
+    // external parameters, which the user can choose while creating the
+    // visualizer bundle
+    stepSize = 1 // the default value of this parameter is zero
+    params = {
+        /** md
+- stepSize: How far to step when the user presses a left or right arrow key.
+         **/
+        stepSize: {
+            value: this.stepSize,
+            forceType: 'integer',
+            displayName: 'Step size',
+            required: true,
+        },
+    }
+
+    // internal parameters. the value of `index` will always be overwritten in
+    // `inhabit()`, but TypeScript can't infer that, so we have to give `index`
+    // an initial value
+    index = 0
+    justStepped = false
+    flash = 0
+
+    checkParameters() {
+        const status = super.checkParameters()
+
+        // make sure the step size is positive
+        if (this.params.stepSize.value <= 0) {
+            status.isValid = false
+            status.errors.push('Step size must be positive')
+        }
+
+        return status
+    }
+
+    inhabit(element: HTMLElement): void {
+        // do the behind-the-scenes work of managing p5 canvases. this work is
+        // finicky, and your visualizer won't work unless it's done properly, so
+        // you should leave it to the superclass unless you have a very
+        // compelling reason to do it yourself
+        super.inhabit(element)
+    }
+
+    setup() {
+        // the base setup function includes the `createCanvas()` call that must
+        // appear in every p5 setup function
+        //
+        //   https://p5js.org/reference/#/p5/createCanvas
+        //
+        super.setup()
+
+        // accessing `this.sketch` triggers some consistency checks, so it's
+        // most parsimonious to access it at most once per function call,
+        // storing its value as a local constant for future reference
+        const sketch = this.sketch
+
+        // set the stroke color and text alignment, which won't change from
+        // frame to frame
+        sketch.stroke(51, 51, 255)
+        sketch.textAlign(sketch.CENTER, sketch.CENTER)
+
```

Comment:
Maybe you could write this bit in setup as
```
this.sketch.stroke(<color>).textAlign(this.sketch.CENTER, this.sketch.CENTER)
```
to show how chaining works in p5. And then it's down to three accesses of this.sketch in setup, which is pretty minimal, especially since setup() is only called once. And so then you could skip the parsimony comment in setup() since you have the exact same comment in draw() and it looks a bit redundant. The fact is it could only matter in draw (not sure how relevant it really even is there, without benchmarking).

Reply:
Oh, and I said "<color>" because (51,51,255) is a bit "magic-number-ish". I'd suggest in the template it's mildly preferable to use a [CSS color string](https://www.w3schools.com/cssref/css_colors.php)

Reply:
I've switched to chaining when doing settings in both `setup()` and `draw()`.

I took the colors from the page style, so I've revised the sketch to grab them automatically during setup.

Reply:
Not that it is any kind of  a big deal, but I am not seeing any chaining in setup(). Just noting the discrepancy with what you said; feel free to resolve that either way. I have commented elsewhere on the grabbing colors thing.